--- Day 14: Parabolic Reflector Dish ---
You reach the place where all of the mirrors were pointing: a massive parabolic reflector dish attached to the side of another large mountain.

The dish is made up of many small mirrors, but while the mirrors themselves are roughly in the shape of a parabolic reflector dish, each individual mirror seems to be pointing in slightly the wrong direction. If the dish is meant to focus light, all it's doing right now is sending it in a vague direction.

This system must be what provides the energy for the lava! If you focus the reflector dish, maybe you can go where it's pointing and use the light to fix the lava production.

Upon closer inspection, the individual mirrors each appear to be connected via an elaborate system of ropes and pulleys to a large metal platform below the dish. The platform is covered in large rocks of various shapes. Depending on their position, the weight of the rocks deforms the platform, and the shape of the platform controls which ropes move and ultimately the focus of the dish.

In short: if you move the rocks, you can focus the dish. The platform even has a control panel on the side that lets you tilt it in one of four directions! The rounded rocks (O) will roll when the platform is tilted, while the cube-shaped rocks (#) will stay in place. You note the positions of all of the empty spaces (.) and rocks (your puzzle input). For example:

O....#....
O.OO#....#
.....##...
OO.#O....O
.O.....O#.
O.#..O.#.#
..O..#O..O
.......O..
#....###..
#OO..#....
Start by tilting the lever so all of the rocks will slide north as far as they will go:

OOOO.#.O..
OO..#....#
OO..O##..O
O..#.OO...
........#.
..#....#.#
..O..#.O.O
..O.......
#....###..
#....#....
You notice that the support beams along the north side of the platform are damaged; to ensure the platform doesn't collapse, you should calculate the total load on the north support beams.

The amount of load caused by a single rounded rock (O) is equal to the number of rows from the rock to the south edge of the platform, including the row the rock is on. (Cube-shaped rocks (#) don't contribute to load.) So, the amount of load caused by each rock in each row is as follows:

OOOO.#.O.. 10
OO..#....#  9
OO..O##..O  8
O..#.OO...  7
........#.  6
..#....#.#  5
..O..#.O.O  4
..O.......  3
#....###..  2
#....#....  1
The total load is the sum of the load caused by all of the rounded rocks. In this example, the total load is 136.

Tilt the platform so that the rounded rocks all roll north. Afterward, what is the total load on the north support beams?

In [45]:
puzzle_data = """.......#...#O.O..#O#...#O#.....#..#O.#..####....OO....#..O.........#..O....#O....#..#O...#.#.#.O.OOO
.O...#O..O#O#OO....#..#....##..OO..###...O#O..#....O........O....O.O.....#.OO...O..OO...#.OO.O#.O#..
OO.#..#O.##O...OO.#..#..OO#.#.#OO.....#..........OO..........#....O#...O...##.O...O...##..O..##.....
....O.O##..O.OO.O....#...O..#.......#.#O#..O.......OO.O...O....#.O..#.#OO#..##.#O#..O..O...#.....#.#
#....O..........#..#.#......#.......O.#..##O..#...#O.#..OO...#..OO.O...OO.O...#...OO...O#..#....O.OO
.....##........##..........O.#..#......OOOOO....#.#O.#..OO#...O#...##OO....##.####.#O.O#O.O##O#...#.
.O..O.#.OOO.O..O.O.O...O.OO......O....O.#.O..#....#.#.....O..O...#...........#.#O#.......#...#.##.O.
...O..#....O###..O.O..O###.#..O..O..O...O..O.O......OOO.O....O##.OOOO..#......#O...#OO...O#.....##..
O.#..O...O...O#..O..#.#....O.O.#..O.....O.....O...O#.O.O..O.O.#O#OOOO#..O..#O.....#..O#..O.O.OO.OO..
#..OO.#...OO###O#...O#OO......#..O.#..O.#.O##.....##OOO.O................OO.O.#O...#..#.....O...OOO.
#O...O.O.....#O###.OO....O#...O....O#......O.OOOO.###....O.#O.#.O.........#.O.....#.#........O......
.O...#..#.O.#.OO.O........#O..OOO.....O.....#O..#.........OO#....OO..###.....OO.O.#O#O...O..#..##..O
O..O...#.#..O.....O#...#........#.#O.#..#.....#O#O.....O......O#......O#........O...OO.OO....#...#..
...##O...###OO..#....OO.##...O..#.#O.O#.##O#.....#.#.O#...#.........OO.O.O#O#..O##....O.OOO#..OO#...
.......O.O.OO.O...OO..#....O#.......O..O..O...#.OO.#..#O...OO...O..OO..#O...##O..O#....#..#.##..OOO.
.O..#O..O..O....##.#.#..OO..OO##....#.....O...#.#.#..#......#.O..#.O#.#.##O#.....#.OO....#O.O..O....
..#OOOOO.....O....O#...#.....#.......OO...#O..O#O...#...#..OO......OO..O.##.#.O.................O.#.
....O..O.......#..##...O.OO.#..##.O...##...#O#..O.##...O#OO...O....#.#.O.#.O#.OO.O#..O.O....#.....#O
#.O....O.O#O.O.O..........O.O#OO..O....O.#O..OO.......O..........O#....#..#....#.#............O##...
..........O.....O.O.#.............O.....O.....O.OO#..#.#......##.....#.#..O...O.O.O..O.O#O.O..O.....
...#..#........#.OO.#.......O.........#.....#.#.#.O....#..O..O.##O..#..O#.O......OO.O#.O#.....#....#
O.....O#.##O.##..O.O.#..O#...#.#.#.O..........OO....O#.#.......#.O.O#.O#O.###.......#O..O#...O....#.
O.##OOO..O.#..OO......O.#.OO...#.O...O..#O.....O...O.#..#O.O...O.....#.O.#.##.O#..........#O#....#..
O#....OOO#.....O...##......O.O.....O.#......O..O#.O.....O..O.O.O.#...O......#...O#.#O........O...OO.
O.....O......O...#..#...#..O..O#.....#O..OOOO#..#..OO......O....#O.#OO##........#...O......O....O.OO
..O#O#.........O..#.#........#.....#..O#..#O.#.O.O##O..O.O....O.#...................O...O....OOO....
OO...O....O.....#O..OO.#..OO...O....#..O..OO#....O..O.O......O..##...#OOO.O...#.O.#..O....O.O...#.O.
#O###...OO.#.....O#....O..OO..#O##......OOO.#OO....O..#O..#.#.O##.#.O.#..#.#OO.#.#.#........O.O.....
O.....O..OO#...#......O..##.O..O.....O....O.......OO......OO.O.##..#..O...O#.O###..OO..OO..O##.O.O.#
#...O...O..........OO#O..#...#....O.#..OO...#...O.O.#.O..##.O.#....O#....O...#..O#...O..O#O..O.O...#
.#..#O#..#O.#.O.....#.#O....#..O..#..O###......#O.#O#.#OO.O..OO.O#...OO..O..#......#.....OO.....#...
.#........O.#...#..O.....OO...#.#OO#......#O..#.........O##..#.....OO...O....#.....#.##....OO.......
...O#.#..#O..#.O.........O..O.#.OO...O....O.O#..OO.OOO..O...#..OO..#O....O....#.O.O......O.#OO#.O...
....#.#.O.O...OO##O..OO.....O...#O...#..#O#O#......#..O.##.O#..O....#...O.OO..O.O#O..O...O.##...##..
.###.OO#...O..O.#.....O.O#.#OO#OOOO..#..O..#OOO.#....OO.O#..........OO......#.#.O...OO.O..#..O..#...
OO.....O.......#........O...###.O......O..##.OO...OO..O...#OO....#.O#..O...#O....O..O.......#.......
...#.....#.O#...O#O.....O....#O........O#..#.#O..#.....#...#.O....#..#..OO.O#........#.....O..###O#.
.O#........#O.O..O.#...O.........#O...O....#...O#..O..##O.#O.#OO.O.....OOOO..#...O..O.O.#O#..O...#.O
....#O.#....#O...#.#.#..#.#....#.#..#..#..O.........#O..#...O....O..##O...O.......##.......#..O.O.##
##OO.O#.O....O.O...O##O.#...O#O..#....#....O......O.......O.....O#...OOO#.....O#..O#......OOO.#.#...
...O.......O..#....OO....O.#O#...OOOO.OO.#..O.##.O..O..#.OO..O#.....O..#O#O.#.#.......#.#..#O....O..
....##O..##.#O.O.OO.......#O##.#...#.O....O.#...#...O...O....#.....##..O..#.....OO.#O....O.O#.OO....
.O...O##...O...#O...#O....O..##....O.###.O...OO.OO.OO.O...O..#..##..#.....O...O...#...O#OO.#...O..O.
#.OO..O.......O.#.OOO.O#.....O##.##OO.##.O...O.O..#.O.#......OO..#O.........O.......#.#..##.#..O....
...##.....#.....O#.#.O.....#..O....#.......OO...#.OO..O.OO..#..#..OO..............O##............OO.
.OOOO.......O..O......##..O#..#....O.OO.#O##O...#.#......OOO.#.O##O..OO..O#....O#...#..#.#O#..#O.O#.
.OOO.#.#O.#O.....#........OOO.#.O...#....O....O.OOO...OO.O.O.O.#.OO.O..#OO..O##.O.##O.##....OO....O.
#O#...#....#O..#......O#O.#O.O.O.##..#O.#O.#..O...#O...OO.O...#OO.O....#OOO..O...O.#OO.O.....O...#..
...O..#.O..#O#.O....O.OO....OOOO..O...O#.....O.#O....O.#O.O.....O.#.#.O......O.......#...#.#O..####.
O..O.##....#......O#.O.O.#.........O.#.....#.....#...O..O...O...O.#..OO.#........OO.##...#.........O
OO#....O..O..OO.O..O.O#OO.#OO..#.....#O...O#......##.#...#....O....#O.....O....OOO...O##OO.O..O..###
.###.O.O#.#....O.OO#O.O........O....OO#....O.#O.O....OO..........OOO.##.....#...OO......O.........O#
.#.O.###O#.O#...OO.O...O..#.O.#...#..OO..#O#...#...O..O......O..O...O.#.#.....#.....#.......#OO.#...
#...#O#.#......O....O..O...O...O#.O.O.#O...###.O##.O...O....O#O..OO..#O#O.#.###..#.#...###O..OO.....
..O#..O.O#O.O#.#...O.OOO#....O.#..#.....O....#.#....O..O.#.O..OOOO#.O#..###.O..O.#.O.O...OOO...O..#.
#.O.O.O.#..O.O.O#.O....O...#O.#.OO...#O....#..O..OO.O#....#.O#......O#O..........#OOO..##...O..O...O
.O..O....O..#..O......##...#O....O....#...#O....OO..O.#O..##..O..............##..OO...#.....#O.###..
##.O...#OO....O.O......OO.#.#..O.##O......#.#.#..#.O...##O.#O......#O..O..#..#...#....O##..#........
.#OOO#....O.#..OO...#..O..#....##..........O.#....##.#.#O...#.O.O...O.O#.O........#..#.O..#......#..
#..O..#..O..#O...OO..#.#......O..OO.O..O...OOO.#.OO..O.#O#O.............O.O.O..O.#.#...OO.....#.....
..O......O.#...##....#O.O.#.O.##...O......O..##..#......#.#....#.O.O..#O.#.O.#...OO....O.O.OO#......
O...O#.....#O..#..OOO.#.....O#OO##.#...O..O.O..##..O..#O.O..O...#..........O.#......O##.#..#..OO.O##
...OO..OOO.#O.O........O.......O.....O.O.O.O#O..O.O...OO..OO.....#.OOO.OO.O#..O#.O.O...##..O..#.....
O.O.##O.#.OO..##...#O.O...#.#......O.....OO.O.#O.O.O....###......OO#O#..OO..O..#.............#....#.
....#.OO....O....#...#...#.#....O#...O..#OOO.O.O.O................#.......#..OO.#..OO.#..O#.....#...
....O#....###.O.#....#OO..##...O...#..#.O.....O..O.###..#....OO#..O...#OO..#...OOO.O..#.O#..O..O###.
..#O.#....OO..#.#.##...O..##.......O...#O...OO#...O#.O..O#O.........OO..O..O.......O.O....#.....O.O.
O...O#O...#OO..#..#.O#O...OO#.#O#..O........O.#....###..O#OO.O....OO.O..O.#.......#.OOO.....##O.....
O##.##O.....OO...O....OO....#....O..#.......#.###...O..O......#......OO#..O#.OO.#....O##.........##O
.OO...#O.##..OOO####.#.........O....#..#.....#O.O..O...#.#..#O..#O...O..O..#O.O..#.........#.O......
#....####OO#OO..O....#.#..#...O#OOO#....#.O..#..#.O.............##O..#O..OO#O.....#.#.O..O#.#..O..O.
OO......O.OOO#O..#..O.O..#.O...O.#..#OO#.O..##...#......OO...O#O..#....##.....#.OO.#..O#.#OO.O.#....
.O##.O....#.....OO#.O#O##..O.....O#.####OO...O....O..O#..#O#..OO....#.O.#.O.#.OOO#O....OO#....#.#O.#
O....#O###..#....O..O.#.#........O..#..OO...........O...OO....#........#O....O#.O..###O#O..#...O#O..
..#.O.#.O....##.O..O#..O#..#...O.O......##O....O#.#.#O.#O.#.O.O....O.#.......O.OO...#...O.OO#..O#..O
.O.O.###...OO...O...#O...#OOO.#..##O..O.O#.....O.O...O..O.....###........#.....O..O...OO#..O.......O
..#O...#..OO.O..O#O.O............#......#.......#...O.O.......OO....#..#..O#O.O......#O#....O.#.#..O
......O........O....O......#.O.#.#..OO.##.O....#........#..O.....O.#.....O..#OO#..O##O.####OO.#....#
.OO.##O....OO.O...OO#O...O#...#.......O.....O#..#.O.##.O.....O.....O..OOO........O..#....O.O#..#.O#.
#.#.#....O....O.O.O..O.O#O.#.#....#O.O.#.O#..OOO.OOOOO.......OO...O..O.....O..#.#.#..O..#O...#O..#..
..O.O....O.O..OOO#O.#O.#.#...O..##...........O.##.#.OO#.O..O.O.....OO..#....#O#..O..#O.OO#.#.##..O.#
..O.O.#O...O..#O...#...O.....O#...#..O..O.......#.....#O#....#....O....#..###OO#O..#.O.O.O.#.#....#O
...O..O.O.....O...O...#..O....#O.#.O#.#.....O#...O..##.O..O..#.#O#.O..O.#.O....O.O.O........O.......
O....O..#...#O..........O##...#O#...#.......O#.O#...##O........O..###..O.O..#..OO..O##OOO..O.O.O.#..
......O..#..#.##O#O..O..#O#.#....O....OO.O.O...O..OO..OO.##...##..........O.#O#.O.O.O##..O#..O.#OOOO
.O...O#O..........OO#.O..O.O.##O#...OO.####.O....###...#.#.O.......O..O.OO..#O.....O....O....OOOO.O.
O..#O......O.#.#..#....O.....O#.#.......#.##O.O.......#..#......#..#...#.....O...O..O...O#.O.#..#...
OO#O..O###..#O#..O.#....#O.......#.O###OO#.....###O#..O........OO#......O#OO.......#.O#......##....O
..O..#...O.#....####....O.OO#..#O........O.#.#OOO#OO..#..#O.....#..#OO..O..O....#...O....O...O...O#.
.O..#..O..#.OO.#.#..O..O......#.O.O..#..#...#...O.#.......O#..#O....OO.#.O.####......O......OOO..#.O
...#..OO..OO..#.O#........OOO.O.O..#O.O..##.#....##O#.O.O.#O.....O#..OOO.##.#.O##.OOO...O..O..O.#O..
.......O........O...O.#....O..#.#.........#......#..O.#..##..OO.....#...##...O.O.O..#..O..OO...#.O..
.....#.....O#O.#.#..O..O#.O....OO......#.#....#.O...O........O.....O....OO.#O...#O..#..O..#...#...OO
...#.O..O....#.#....##.#OOO....#.O.....O.....O.##.###.#...#.O....#....O....#.#O#O.O..O.....##O.#....
.OO...#....#O.#.O..O........#.O#.#.................#.#.##......##.#.O.....O##....O......O...........
O........#O..#...#....#O.....#......#....O..O....O..#....O#O.....O#....OO.OO..##.#.#...#.#.O...#.#.#
O..#.....O.O....O#O..OO##O...O..#.##O##....O.O.O....#............#.....O#.##..#O....O....O.O..O.....
.........#..O.O.O#......O...........OO..#.......O#........O..#..O.#.#.#...O.......O.O...O..O......O.
.###..OO#.....O##.O..O#O#...#..#...#...#O#O.....##............O.....#.O##..O#.....OO..##O....#.#....
O#....OO.O#....#...O.#..#.....#.........OO..O#OO.#O......#...OOO.##...O.OOOO.O.....O#OO....O.#...O.."""

test_data = """O....#....
O.OO#....#
.....##...
OO.#O....O
.O.....O#.
O.#..O.#.#
..O..#O..O
.......O..
#....###..
#OO..#...."""

In [40]:
import numpy as np

def transform_input_to_matrix(data):
    
    return np.array([[char for char in line] for line in data.splitlines()])

def rolle_north(matrix):
    
    def rolle_west(list_of_rocks):
    
        left_pointer = 0

        for right_pointer in range(len(list_of_rocks)):

            # if no stone found push left pointer
            if list_of_rocks[right_pointer] == '.':
                continue

            # if square stone found, set left pointer to the position behinde that
            elif list_of_rocks[right_pointer] == '#':
                if right_pointer + 1 < len(list_of_rocks):
                    left_pointer = right_pointer + 1

            # if round stone found, swap place with left pointer
            elif list_of_rocks[right_pointer] == 'O':
                list_of_rocks[right_pointer] = '.'
                list_of_rocks[left_pointer] = 'O'

                left_pointer += 1

        return list_of_rocks
    
    matrix_t = matrix.transpose()
    
    for row_index in range(len(matrix_t)):
        
        matrix_t[row_index] = rolle_west(matrix_t[row_index])
        
    return matrix_t.transpose()

def get_weight(matrix):
    
    res = 0
    
    for index, row in enumerate(matrix):
        res += sum(element == 'O' for element in row) * (len(matrix) - index)
    
    return res

In [46]:
def solve_puzzle(input_data):
    
    matrix = transform_input_to_matrix(input_data)
    stones_shifted_matrix = rolle_north(matrix)
    
    return get_weight(stones_shifted_matrix)
    
solve_puzzle(puzzle_data)

109385

109385 is correct

--- Part Two ---
The parabolic reflector dish deforms, but not in a way that focuses the beam. To do that, you'll need to move the rocks to the edges of the platform. Fortunately, a button on the side of the control panel labeled "spin cycle" attempts to do just that!

Each cycle tilts the platform four times so that the rounded rocks roll north, then west, then south, then east. After each tilt, the rounded rocks roll as far as they can before the platform tilts in the next direction. After one cycle, the platform will have finished rolling the rounded rocks in those four directions in that order.

Here's what happens in the example above after each of the first few cycles:

After 1 cycle:
.....#....
....#...O#
...OO##...
.OO#......
.....OOO#.
.O#...O#.#
....O#....
......OOOO
#...O###..
#..OO#....

After 2 cycles:
.....#....
....#...O#
.....##...
..O#......
.....OOO#.
.O#...O#.#
....O#...O
.......OOO
#..OO###..
#.OOO#...O

After 3 cycles:
.....#....
....#...O#
.....##...
..O#......
.....OOO#.
.O#...O#.#
....O#...O
.......OOO
#...O###.O
#.OOO#...O
This process should work if you leave it running long enough, but you're still worried about the north support beams. To make sure they'll survive for a while, you need to calculate the total load on the north support beams after 1000000000 cycles.

In the above example, after 1000000000 cycles, the total load on the north support beams is 64.

Run the spin cycle for 1000000000 cycles. Afterward, what is the total load on the north support beams?

In [ ]:
def run_cycle(matrix):
    

In [ ]:
def solve_second_puzzle(input_data):
    
    matrix = transform_input_to_matrix(input_data)
    
    for i in range(1000000000):
        
        stones_shifted_matrix = rolle_north(matrix)
    
    return get_weight(stones_shifted_matrix)
    
solve_second_puzzle(puzzle_data)